src/transform/src/redundant_join.rs

```diff
@@ -103,9 +107,31 @@ impl RedundantJoin {
                     Ok(result)
                 }
 
-                MirRelationExpr::LetRec { .. } => {
-                    // TODO
-                    Err(crate::TransformError::LetRecUnsupported)?
+                MirRelationExpr::LetRec { ids, values, body } => {
+                    // As a first approximation, we extend the `lets` context as follows.
+                    for (id, value) in zip_eq(ids.iter(), values.iter()) {
+                        let val_type = value.typ();
+                        let prov_new = ProvInfo::make_leaf(Id::Local(*id), val_type.arity());
+                        let prov_old = lets.insert(Id::Local(*id), vec![prov_new]);
+                        assert!(prov_old.is_none());
+                    }
+
+                    // In other words, we don't attempt to derive additional
+                    // provenance information for a binding from its `value`.
+                    //
+                    // We descend into the values and the body with the naively
+                    // extended context.
+                    for value in values.iter_mut() {
+                        self.action(value, lets)?;
```

Comment:
How about adding the result of the `action` to `lets` the same way as the `Let` case does? `Get`s that reach across iteration boundaries would still be leaves, but Gets that refer to the same iteration would have some provenance that they collected within the iteration. I think this would be equivalent to doing exactly one step of a pessimistic fixpoint loop. (`ProvInfo::make_leaf` is the `top`, right?)

```
                    for (id, value) in zip_eq(ids.iter(), values.iter_mut()) {
                        let value_prov = self.action(value, lets)?;
                        let prov_old = lets.insert(Id::Local(*id), value_prov);
                        assert!(prov_old.is_none());
                    }
```

Reply:
The problem is that I'm not sure that the `Vec<ProvInfo>` inference is monotonic and I don't want to spend the cycles to ensure that it is at the moment.

Reply:
Ok!